File: solutions/cpp/k-th_node_binary_tree.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>
#include <stdexcept>

using std::cout;
using std::endl;
using std::exception;
using std::length_error;
using std::unique_ptr;

template <typename T>
struct BinaryTreeNode {
  T data;
  unique_ptr<BinaryTreeNode<T>> left, right;
  int size;
};

// @include
const BinaryTreeNode<int>* FindKthNodeBinaryTree(
    const unique_ptr<BinaryTreeNode<int>>& root, int k) {
  const auto* n = root.get();
  while (n) {
    int left_size = n->left ? n->left->size : 0;
    if (left_size < k - 1) {
      k -= (left_size + 1);
      n = n->right.get();
    } else if (left_size == k - 1) {
      return n;
    } else {  // left_size > k - 1.
      n = n->left.get();
    }
  }
  throw length_error("no k-th node in binary tree");
}
// @exclude

int main(int argc, char* argv[]) {
  //  size field
  //      6
  //    2   3
  //  1    1 1
  //
  //  data field
  //      3
  //    2   5
  //  1    4 6
  auto root = unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  root->size = 6;
  root->data = 3;
  root->left = unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  root->left->size = 2;
  root->left->data = 2;
  root->left->left = unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  root->left->left->size = 1;
  root->left->left->data = 1;
  root->right = unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  root->right->size = 3;
  root->right->data = 5;
  root->right->left = unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  root->right->left->size = 1;
  root->right->left->data = 4;
  root->right->right = unique_ptr<BinaryTreeNode<int>>(new BinaryTreeNode<int>());
  root->right->right->size = 1;
  root->right->right->data = 6;
  // should throw
  try {
    FindKthNodeBinaryTree(root, 0);
  }
  catch (const exception& e) {
    cout << e.what() << endl;
  }
  // should output 1
  assert(FindKthNodeBinaryTree(root, 1)->data == 1);
  cout << (FindKthNodeBinaryTree(root, 1))->data << endl;
  // should output 2
  assert(FindKthNodeBinaryTree(root, 2)->data == 2);
  cout << (FindKthNodeBinaryTree(root, 2))->data << endl;
  // should output 3
  assert(FindKthNodeBinaryTree(root, 3)->data == 3);
  cout << (FindKthNodeBinaryTree(root, 3))->data << endl;
  // should output 4
  assert(FindKthNodeBinaryTree(root, 4)->data == 4);
  cout << (FindKthNodeBinaryTree(root, 4))->data << endl;
  // should output 5
  assert(FindKthNodeBinaryTree(root, 5)->data == 5);
  cout << (FindKthNodeBinaryTree(root, 5))->data << endl;
  // should output 6
  assert(FindKthNodeBinaryTree(root, 6)->data == 6);
  cout << (FindKthNodeBinaryTree(root, 6))->data << endl;
  // should throw
  try {
    FindKthNodeBinaryTree(root, 7);
  }
  catch (const exception& e) {
    cout << e.what() << endl;
  }
  return 0;
}

File: solutions/python/k_th_node_binary_tree.py
class BinaryTreeNode:

    def __init__(self, data=None, left=None, right=None, size=None):
        self.data = data
        self.left = left
        self.right = right
        self.size = size


# @include
def find_kth_node_binary_tree(tree, k):
    while tree:
        left_size = tree.left.size if tree.left else 0
        if left_size + 1 < k:  # k-th node must be in right subtree of tree.
            k -= left_size + 1
            tree = tree.right
        elif left_size == k - 1:  # k-th is iter itself.
            return tree
        else:  # k-th node must be in left subtree of iter.
            tree = tree.left
    return None  # If k is between 1 and the tree size, this is unreachable.
# @exclude


def main():
    #  size field
    #      6
    #    2   3
    #  1    1 1
    #
    #  data field
    #      3
    #    2   5
    #  1    4 6
    root = BinaryTreeNode()
    root.size = 6
    root.data = 3
    root.left = BinaryTreeNode()
    root.left.size = 2
    root.left.data = 2
    root.left.left = BinaryTreeNode()
    root.left.left.size = 1
    root.left.left.data = 1
    root.right = BinaryTreeNode()
    root.right.size = 3
    root.right.data = 5
    root.right.left = BinaryTreeNode()
    root.right.left.size = 1
    root.right.left.data = 4
    root.right.right = BinaryTreeNode()
    root.right.right.size = 1
    root.right.right.data = 6
    assert None is find_kth_node_binary_tree(root, 0)
    # should output 1
    assert find_kth_node_binary_tree(root, 1).data == 1
    print((find_kth_node_binary_tree(root, 1)).data)
    # should output 2
    assert find_kth_node_binary_tree(root, 2).data == 2
    print((find_kth_node_binary_tree(root, 2)).data)
    # should output 3
    assert find_kth_node_binary_tree(root, 3).data == 3
    print((find_kth_node_binary_tree(root, 3)).data)
    # should output 4
    assert find_kth_node_binary_tree(root, 4).data == 4
    print((find_kth_node_binary_tree(root, 4)).data)
    # should output 5
    assert find_kth_node_binary_tree(root, 5).data == 5
    print((find_kth_node_binary_tree(root, 5)).data)
    # should output 6
    assert find_kth_node_binary_tree(root, 6).data == 6
    print((find_kth_node_binary_tree(root, 6)).data)
    assert None is find_kth_node_binary_tree(root, 7)


if __name__ == '__main__':
    main()
